doc/developer/design/20230421_stabilize_monotonic_select.md

```diff
@@ -0,0 +1,149 @@
+
+The MVP PR gated the execution of monotonic one-shot `SELECT`s behind the feature flag `enable_monotonic_oneshot_selects`, turned off by default. The goal of this design is to outline the necessary steps to stabilize this feature and retire the feature flag.
+
+# Concepts and Challenges
+[explanation]: #explanation
+
+This section explains relevant concepts for monotonic one-shot `SELECT`s. We focus here on important background and challenges to stabilization, while the [Reference Explanation](#reference-explanation) section introduces proposed solution approaches to the challenges identified.
+
+## Logical vs. Physical Monotonicity
+
+Sources can classified into monotonic, i.e., present no retractions to Materialize, or non-monotonic. A monotonic source exhibits a property that we refer to as _physical monotonicity_, namely that the stream of updates produced by the source is retraction-free. One could also consider a relaxed property, which we term _logical monotonicity_. In logical monotonicity, the result of applying a `consolidate_named` operator to the stream is retraction-free, i.e., the stream may contain both additions and retractions for the same row in the same timestamp, but the consolidated view of the stream is retraction-free at every timestamp.
+
+## One-Shot `SELECT`s: Almost Monotonic
+
+When planning a one-shot `SELECT`, a `DataflowDescription` is created. A `DataflowDescription` contains both `as_of` and `until` frontiers. The interpretation of these frontiers is that when reading data from a source, a dataflow will fast-forward updates less than the `as_of` to the `as_of` frontier so as to construct a logical snapshot of the input. Then, further updates available in the input will be streamed continuously; however, any updates that are not less than the `until` will be discarded. When the coordinator creates a peek plan, an `as_of` value is provided. Should the planning not hit a fast path, then the `until` value of the `DataflowDescription` will be set to a single-coordinate timestamp conceptually equivalent to `as_of + 1`. This behavior ensures that the input streams produced, e.g., by `persist_source`, will comply with logical, but not physical, monotonicity.
+
+## MVP of the Feature
+
+The MVP PR MaterializeInc/materialize#18546 exploited the fact that one-shot `SELECT`s are almost monotonic, i.e., already exhibit logical monotonicity, to improve their plans. Each `Plan` LIR operator that has a monotonic variant, namely hierarchical reduction and top-k, is given the possibility to convert a non-monotonic operator `Plan` into a monotonic one with forced consolidation, represented by a `must_consolidate` flag. The rendering of these LIR operators then respects the `must_consolidate` flag by either introducing a `consolidate_named` or not on the key-value input obtained after computing group keys and relevant values.
+
+## Relationship to Monotonicity Analysis in MIR and `EXPLAIN`
```

Comment:
nit: we also need to ensure that in `EXPLAIN OPTIMIZER TRACE` we records the different steps (right after `mir_to_lir` lowering and then after each `lir` pass.